distributed/client.py

```diff
@@ -2600,6 +2600,59 @@ def __anext__(self):
 
     next = __next__
 
+    def next_batch(self, block=True):
+        """ Get next batch of futures from as_completed iterator
+
+        Parameters
+        ----------
+        block: bool, optional
+            If True then wait until we have some result, otherwise return
+            immediately, even with an empty list.  Defaults to True.
+
+        Examples
+        --------
+        >>> ac = as_completed(futures)  # doctest: +SKIP
+        >>> client.gather(ac.next_batch())  # doctest: +SKIP
+        [4, 1, 3]
+
+        >>> client.gather(ac.next_batch(block=False))  # doctest: +SKIP
+        []
+
+        Returns
+        -------
+        List of futures or (future, result) tuples
+        """
+        if block:
+            batch = [next(self)]
+        else:
+            batch = []
+        with self.lock:
```

Comment:
I suppose general paranoia.  You're right though.  I think we've used the lock historically because we also check `self.futures` which is not threadsafe and is modified in a separate thread.  That's not the case in this code though (except implicitly in `next`) .  Removing.